### PR TITLE
[Feature] Max Value Writer

### PR DESCRIPTION
**Pull request by albertbou92**
## Description

A Writer class for composable replay buffers that keeps the top elements based on some ranking key.

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: torchrl/data/replay_buffers/writers.py [hunk 1]
```diff
@@ -92,3 +93,68 @@ def extend(self, data: Sequence) -> torch.Tensor:
         data["index"] = index
         self._storage[index] = data
         return index
+
+
+class TensorDictMaxValueWriter(Writer):
+    """A Writer class for composable replay buffers that keeps the top elements based on some ranking key.
+
+    If rank_key is not provided, the key will be ("next", "reward").
+    """
```
Comment (vmoens) [inline torchrl/data/replay_buffers/writers.py:102]: Let's add an example here!

File: torchrl/data/replay_buffers/writers.py [hunk 2]
```diff
@@ -92,3 +93,68 @@ def extend(self, data: Sequence) -> torch.Tensor:
...
+        self._rank_key = rank_key
+        if self._rank_key is None:
+            self._rank_key = ("next", "reward")
+
+    def add(self, data: Any) -> int:
+
+        ret = None
+
+        # Sum the rank key, in case it is a whole trajectory
+        rank_data = data.get("_data")[self._rank_key].sum()
```
Comment (vmoens) [inline torchrl/data/replay_buffers/writers.py:117]:
```suggestion
        rank_data = data.get("_data", self._rank_key).sum()
```
We need to make sure that this works with tensorclass too

File: torchrl/data/replay_buffers/writers.py [hunk 3]
```diff
@@ -92,3 +93,68 @@ def extend(self, data: Sequence) -> torch.Tensor:
...
+
+        if rank_data is None:
+            raise ValueError(f"Rank key {self._rank_key} not found in data.")
+
+        # If the buffer is not full, add the data
+        if len(self._storage) < self._storage.max_size:
+
+            ret = self._cursor
+            data["index"] = ret
```
Comment (vmoens) [inline torchrl/data/replay_buffers/writers.py:126]:
```suggestion
            data.set("index", ret)
```

File: torchrl/data/replay_buffers/writers.py [hunk 4]
```diff
@@ -92,3 +93,68 @@ def extend(self, data: Sequence) -> torch.Tensor:
...
+
+            # retrieve position of the smallest value
+            min_sample = heapq.heappop(self._current_top_values)
+            min_sample_value = min_sample[1]
+
+            # replace the smallest value with the new value
+            self._storage[min_sample_value] = data
+
+            # set new data index
+            data["index"] = min_sample_value
```
Comment (vmoens) [inline torchrl/data/replay_buffers/writers.py:144]:
```suggestion
            data.set("index", min_sample_value)
```

File: torchrl/data/replay_buffers/writers.py [hunk 5]
```diff
@@ -92,3 +93,68 @@ def extend(self, data: Sequence) -> torch.Tensor:
...
+            ret = min_sample_value
+
+            # Add new reward to the heap
+            heapq.heappush(self._current_top_values, (rank_data, ret))
+
+        return ret
+
+    def extend(self, data: Sequence) -> None:
+        for sample in data:
+            self.add(sample)
```
Comment (vmoens) [inline torchrl/data/replay_buffers/writers.py:156]: Is this efficient when data is a TensorDict?
Shouldn't we batch these ops?

**Review by vmoens — COMMENTED** [review]
Looks promising thanks a lot for this
Can we add it to the docs?

File: torchrl/data/replay_buffers/writers.py [hunk 6]
```diff
@@ -92,3 +93,113 @@ def extend(self, data: Sequence) -> torch.Tensor:
         data["index"] = index
         self._storage[index] = data
         return index
+
+
+class TensorDictMaxValueWriter(Writer):
+    """A Writer class for composable replay buffers that keeps the top elements based on some ranking key.
+
+    If rank_key is not provided, the key will be ("next", "reward").
```
Comment (vmoens) [inline torchrl/data/replay_buffers/writers.py:101]:
```suggestion
    If rank_key is not provided, the key will be ``("next", "reward")``.
```

File: torchrl/data/replay_buffers/writers.py [hunk 7]
```diff
@@ -92,3 +93,113 @@ def extend(self, data: Sequence) -> torch.Tensor:
...
+
+    Examples:
+    >>> import torch
+    >>> from tensordict import TensorDict
+    >>> from torchrl.data import LazyTensorStorage, TensorDictReplayBuffer, TensorDictMaxValueWriter
+    >>> from torchrl.data.replay_buffers.samplers import SamplerWithoutReplacement
+
```
Comment (vmoens) [inline torchrl/data/replay_buffers/writers.py:108]:
```suggestion
```

File: torchrl/data/replay_buffers/writers.py [hunk 8]
```diff
@@ -92,3 +93,113 @@ def extend(self, data: Sequence) -> torch.Tensor:
...
+    9
+
+    >>> td = TensorDict({
+    ...     "key": torch.tensor(range(10, 20)),
+    ...     "obs": torch.tensor(range(10, 20))
+    ... }, batch_size=10)
+    >>> rb.extend(td)
+    >>> print(rb.sample().get("obs").item())
+    19
+
```
Comment (vmoens) [inline torchrl/data/replay_buffers/writers.py:130]:
```suggestion
```

File: torchrl/data/replay_buffers/writers.py [hunk 9]
```diff
@@ -92,3 +93,113 @@ def extend(self, data: Sequence) -> torch.Tensor:
...
+    ...     writer=TensorDictMaxValueWriter(rank_key="key"),
+    ... )
+    >>> td = TensorDict({
+    ...     "key": torch.tensor(range(10)),
+    ...     "obs": torch.tensor(range(10))
+    ... }, batch_size=10)
+    >>> rb.extend(td)
+    >>> print(rb.sample().get("obs").item())
+    9
+
```
Comment (vmoens) [inline torchrl/data/replay_buffers/writers.py:122]:
```suggestion
```

File: torchrl/data/replay_buffers/writers.py [hunk 10]
```diff
@@ -92,3 +93,113 @@ def extend(self, data: Sequence) -> torch.Tensor:
...
+    ...     "key": torch.tensor(range(10)),
+    ...     "obs": torch.tensor(range(10))
+    ... }, batch_size=10)
+    >>> rb.extend(td)
+    >>> print(rb.sample().get("obs").item())
+    19
+    """
+
+    def __init__(self, rank_key=None, **kw) -> None:
+        super().__init__(**kw)
```
Comment (vmoens) [inline torchrl/data/replay_buffers/writers.py:141]:
```suggestion
    def __init__(self, rank_key=None, **kwargs) -> None:
        super().__init__(**kwargs)
```

File: torchrl/data/replay_buffers/writers.py [hunk 11]
```diff
@@ -92,3 +93,113 @@ def extend(self, data: Sequence) -> torch.Tensor:
...
+        self._cursor = 0
+        self._current_top_values = []
+        self._rank_key = rank_key
+        if self._rank_key is None:
+            self._rank_key = ("next", "reward")
+
+    def get_insert_index(self, data: Any) -> int:
+        """Returns the index where the data should be inserted, or None if it should not be inserted."""
+        ret = None
+        rank_data = data.get("_data").get(self._rank_key)
```
Comment (vmoens) [inline torchrl/data/replay_buffers/writers.py:151]:
```suggestion
        rank_data = data.get(("_data", self._rank_key))
```

File: torchrl/data/replay_buffers/writers.py [hunk 12]
```diff
@@ -92,3 +93,113 @@ def extend(self, data: Sequence) -> torch.Tensor:
...
+
+        # Sum the rank key, in case it is a whole trajectory
+        rank_data = rank_data.sum().item()
```
Comment (vmoens) [inline torchrl/data/replay_buffers/writers.py:154]: Is this safe?
Maybe we should document what are the expected shapes for this class, eg
```python
[B, T]
```
but not
```
[B1, B2, T]
```

Another option is to check the number of dimensions of the ranking key OR the name of the last dim of the input tensordict (which should be "time").

Not raising any exception and just doing a plain sum could lead to surprising results I think

File: torchrl/data/replay_buffers/writers.py [hunk 13]
```diff
@@ -92,3 +93,113 @@ def extend(self, data: Sequence) -> torch.Tensor:
...
+
+        if rank_data is None:
+            raise ValueError(f"Rank key {self._rank_key} not found in data.")
```
Comment (vmoens) [inline torchrl/data/replay_buffers/writers.py:157]:
```suggestion
            raise KeyError(f"Rank key {self._rank_key} not found in data.")
```

File: torchrl/data/replay_buffers/writers.py [hunk 14]
```diff
@@ -92,3 +93,113 @@ def extend(self, data: Sequence) -> torch.Tensor:
...
+            index = self.get_insert_index(sample)
+            if index is not None:
+                data_to_replace[index] = i
+
+        # Replace the data in the storage all at once
+        keys = list(data_to_replace.keys())
+        if len(keys) > 0:
+            values = list(data_to_replace.values())
+            data.get("index")[values].copy_(torch.tensor(keys))
+            self._storage[keys] = data[values]
```
Comment (vmoens) [inline torchrl/data/replay_buffers/writers.py:201]:
```suggestion
        keys, values = zip(*data_to_replace.items())
        if len(keys) > 0:
            index = data.get("index")
            index[values] = torch.tensor(keys)
            data.set("index", index)
            self._storage[list(keys)] = data[values]
```

**Review by vmoens — APPROVED** [review]
LGTM thanks for this! I left some suggestions

File: torchrl/data/replay_buffers/writers.py (hunk 12 above)
Comment (albertbou92) [reply torchrl/data/replay_buffers/writers.py]: I adde the first option. Since the ranking values has to be a single value we only allow data of the shape [] and [T] for the `add` method and [B] and [B, T] for the `extend` method. If data had a time dimension, we sum along it. If more than too many dimensions are provided, an exception is raised.

I did not go for checking the dimension names because in examples often the time dimension is not properly named and seemed to restrictive.

File: torchrl/data/replay_buffers/writers.py [hunk 15]
```diff
@@ -92,3 +93,128 @@ def extend(self, data: Sequence) -> torch.Tensor:
...
+    def __init__(self, rank_key=None, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self._cursor = 0
+        self._current_top_values = []
+        self._rank_key = rank_key
+        if self._rank_key is None:
+            self._rank_key = ("next", "reward")
+
+    def get_insert_index(self, data: Any) -> int:
+        """Returns the index where the data should be inserted, or None if it should not be inserted."""
```
Comment (vmoens) [inline torchrl/data/replay_buffers/writers.py:146]:
```suggestion
        """Returns the index where the data should be inserted, or ``None`` if it should not be inserted."""
```

File: torchrl/data/replay_buffers/writers.py [hunk 16]
```diff
@@ -92,3 +93,128 @@ def extend(self, data: Sequence) -> torch.Tensor:
...
+                data_to_replace[index] = i
+
+        # Replace the data in the storage all at once
+        keys, values = zip(*data_to_replace.items())
+        keys, values = list(keys), list(values)
+        if len(keys) > 0:
+            index = data.get("index")
+            index[values] = torch.tensor(keys, dtype=index.dtype)
+            data.set("index", index)
+            self._storage[keys] = data[values]
```
Comment (vmoens) [inline torchrl/data/replay_buffers/writers.py:216]:
```suggestion
        keys, values = zip(*data_to_replace.items())
        if len(keys) > 0:
            index = data.get("index")
            values = list(values)
            keys = index[values] = torch.tensor(keys, dtype=index.dtype)
            data.set("index", index)
            self._storage[keys] = data[values]
```

**Review by vmoens — APPROVED** [review]
LGTM thanks so much!

File: torchrl/data/replay_buffers/writers.py (hunk 12 above)
Comment (vmoens) [reply torchrl/data/replay_buffers/writers.py]: Not always but mostly
if you get your data from env.rollout or collector, it will.
If from there you store the data in a rb, it will keep the tag.
But if you reshape or do other stuff it could go away.
